File: examples/basic_linear.rs
use non_empty_slice::NonEmptyVec;
/// Basic linear spectrogram example
///
/// This example demonstrates:
/// - Creating a simple sine wave
/// - Computing a linear-frequency power spectrogram
/// - Accessing spectrogram data and axes
use spectrograms::{LinearPowerSpectrogram, SpectrogramParams, StftParams, WindowType, nzu};
use std::f64::consts::PI;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Generate a 1-second sine wave at 440 Hz
    let sample_rate = 16000.0;
    let duration = 1.0;
    let frequency = 440.0;

    let samples: Vec<f64> = (0..(duration * sample_rate) as usize)
        .map(|i| (2.0 * PI * frequency * i as f64 / sample_rate).sin())
        .collect();
    let samples = NonEmptyVec::new(samples).unwrap();

    println!("Generated {} samples at {} Hz", samples.len(), sample_rate);

    // Set up spectrogram parameters
    let stft = StftParams::new(
        nzu!(512),           // n_fft: FFT window size
        nzu!(256),           // hop_size: samples between frames
        WindowType::Hanning, // window function
        true,                // centre: pad signal for centered frames
    )?;

    let params = SpectrogramParams::new(stft, sample_rate)?;

    // Compute the spectrogram (this is the convenience API)
    let spec = LinearPowerSpectrogram::compute(&samples, &params, None)?;

    // Access spectrogram properties
    println!("\nSpectrogram properties:");
    println!("  Frequency bins: {}", spec.n_bins());
    println!("  Time frames: {}", spec.n_frames());
    println!("  Shape: {} x {}", spec.n_bins(), spec.n_frames());

    // Access frequency and time axes
    let (f_min, f_max) = spec.axes().frequency_range();
    println!("\nFrequency range: {:.1} Hz to {:.1} Hz", f_min, f_max);

    let duration_computed = spec.axes().duration();
    println!("Duration: {:.3} seconds", duration_computed);

    // Find the peak frequency in the first frame
    let first_frame_data = spec.data().column(0);
    let (peak_bin, peak_power) = first_frame_data
        .iter()
        .enumerate()
        .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
        .unwrap();

    let peak_freq = spec.axes().frequencies()[peak_bin];
    println!(
        "\nPeak in first frame: {:.1} Hz (power: {:.2e})",
        peak_freq, peak_power
    );

    Ok(())
}

File: python/examples/basic_linear.py
#!/usr/bin/env python3
"""
Basic Linear Spectrogram Example

Demonstrates computing a simple linear-frequency power spectrogram.
"""

import numpy as np
import spectrograms as sg


def main():
    print("=" * 60)
    print("Basic Linear Spectrogram Example")
    print("=" * 60)

    # Generate a test signal: 1 second of 440 Hz (A4 note)
    sample_rate = 16000
    duration = 1.0
    frequency = 440.0

    t = np.linspace(0, duration, int(sample_rate * duration), dtype=np.float64)
    samples = np.sin(2 * np.pi * frequency * t)

    print(f"\nGenerated test signal:")
    print(f"  Sample rate: {sample_rate} Hz")
    print(f"  Duration: {duration} s")
    print(f"  Frequency: {frequency} Hz")
    print(f"  Samples: {len(samples)}")

    # Configure STFT parameters
    stft = sg.StftParams(
        n_fft=512,  # FFT size
        hop_size=256,  # Hop between frames
        window="hanning",  # Window function
        centre=True,  # Center frames with padding
    )

    params = sg.SpectrogramParams(stft, sample_rate=sample_rate)

    print(f"\nSTFT parameters:")
    print(f"  FFT size: {stft.n_fft}")
    print(f"  Hop size: {stft.hop_size}")
    print(f"  Window: {stft.window}")
    print(f"  Centered: {stft.centre}")

    # Compute power spectrogram
    print("\nComputing linear power spectrogram...")
    spec = sg.compute_linear_power_spectrogram(samples, params)

    print(f"\nSpectrogram result:")
    print(f"  Shape: {spec.shape}")
    print(f"  Frequency bins: {spec.n_bins}")
    print(f"  Time frames: {spec.n_frames}")
    print(
        f"  Frequency range: {spec.frequency_range()[0]:.1f} - {spec.frequency_range()[1]:.1f} Hz"
    )
    print(f"  Duration: {spec.duration():.3f} s")
    print(f"  Data type: {spec.data.dtype}")
    print(f"  Data shape: {spec.data.shape}")

    # Show some frequency values
    print(f"\nFirst 5 frequency bins:")
    for i in range(min(5, len(spec.frequencies))):
        print(f"  Bin {i}: {spec.frequencies[i]:.2f} Hz")

    # Show some time values
    print(f"\nFirst 5 time frames:")
    for i in range(min(5, len(spec.times))):
        print(f"  Frame {i}: {spec.times[i]:.4f} s")

    # Find the peak frequency (should be close to 440 Hz)
    # Average power across time
    avg_power = np.mean(spec.data, axis=1)
    peak_bin = np.argmax(avg_power)
    peak_freq = spec.frequencies[peak_bin]

    print(f"\nPeak frequency detected:")
    print(f"  Bin {peak_bin}: {peak_freq:.2f} Hz")
    print(f"  Expected: {frequency:.2f} Hz")
    print(f"  Error: {abs(peak_freq - frequency):.2f} Hz")

    print("\nBasic linear spectrogram example completed!")


if __name__ == "__main__":
    main()
